wasm/tests/test.c

```diff
@@ -1724,6 +1724,209 @@ void test_object(void)
+    test("object/remove (non-object first operand)", opa_value_compare(builtin_object_remove(opa_string_terminated("a"),  opa_object()), NULL) == 0);
+
+    opa_set_t *set_keys3 = opa_cast_set(opa_set());
+    opa_set_add(set_keys3, opa_string_terminated("foo"));
+    test("object/remove (key does not exist)", opa_value_compare(builtin_object_remove(&obj3->hdr, &set_keys3->hdr), &obj3->hdr) == 0);
+
+    test("object/remove (second operand not object/set/array)", opa_value_compare(builtin_object_remove(&obj3->hdr, opa_string_terminated("a")), &obj3->hdr) == 0);
+}
+
+void test_object_union(void)
+{
+    test("object/union (both empty)", opa_value_compare(builtin_object_union(opa_object(), opa_object()), opa_object()) == 0);
+
+    opa_object_t *obj1 = opa_cast_object(opa_object());
+    opa_object_insert(obj1, opa_string_terminated("a"), opa_number_int(1));
+    test("object/union (left empty)", opa_value_compare(builtin_object_union(opa_object(), &obj1->hdr), &obj1->hdr) == 0);
+
+    test("object/union (right empty)", opa_value_compare(builtin_object_union(&obj1->hdr, opa_object()), &obj1->hdr) == 0);
+
+    opa_object_t *obj2 = opa_cast_object(opa_object());
+    opa_object_insert(obj2, opa_string_terminated("b"), opa_number_int(2));
+
+    opa_object_t *expected1 = opa_cast_object(opa_object());
+    opa_object_insert(expected1, opa_string_terminated("a"), opa_number_int(1));
+    opa_object_insert(expected1, opa_string_terminated("b"), opa_number_int(2));
+
+    test("object/union (base)", opa_value_compare(builtin_object_union(&obj1->hdr, &obj2->hdr), &expected1->hdr) == 0);
+
+    opa_object_t *o = opa_cast_object(opa_object());
+    opa_object_insert(o, opa_string_terminated("c"), opa_number_int(3));
+    opa_object_t *o2 = opa_cast_object(opa_object());
+    opa_object_insert(o2, opa_string_terminated("b"), &o->hdr);
+    opa_object_t *obj3 = opa_cast_object(opa_object());
+    opa_object_insert(obj3, opa_string_terminated("a"), &o2->hdr);
+
+    opa_object_t *expected2 = opa_cast_object(opa_object());
+    opa_object_insert(expected2, opa_string_terminated("a"),&o2->hdr);
+    opa_object_insert(expected2, opa_string_terminated("b"), opa_number_int(2));
+
+    test("object/union (nested)", opa_value_compare(builtin_object_union(&obj3->hdr, &obj2->hdr), &expected2->hdr) == 0);
+    test("object/union (nested reverse)", opa_value_compare(builtin_object_union(&obj2->hdr, &obj3->hdr), &expected2->hdr) == 0);
+
+    opa_object_t *obj4 = opa_cast_object(opa_object());
+    opa_object_insert(obj4, opa_string_terminated("a"), opa_number_int(2));
+
+    test("object/union (conflict simple)", opa_value_compare(builtin_object_union(&obj1->hdr, &obj4->hdr), &obj4->hdr) == 0);
+
+    opa_object_insert(obj3, opa_string_terminated("d"), opa_number_int(7));
+
+    test("object/union (conflict nested and extra field)", opa_value_compare(builtin_object_union(&obj1->hdr, &obj3->hdr), &obj3->hdr) == 0);
+
+    // Operand 1 -> {"a": {"b": {"c": 1}}, "e": 1}
```

Comment:
💭 We could really use some test helpers...

Reply:
Agreed. I will try to look into this with the next set of builtin implementation changes.